src/drivers/display/fb/lv_linux_fbdev.c

```diff
@@ -275,51 +276,62 @@ static void flush_cb(lv_display_t * disp, const lv_area_t * area, uint8_t * colo
-            dsc->rotated_buf = realloc(dsc->rotated_buf, buf_size);
-            dsc->rotated_buf_size = buf_size;
+    if(rotation != LV_DISPLAY_ROTATION_0) {
+        int32_t src_w;
+        int32_t src_h;
+        uint32_t src_stride;
+
+        /* Direct render mode rotation only works if we rotate the whole screen at the same time
+         * To do that, we use the display's resolution and as the area
+         *  we also grab the current draw buffer so that we can rotate the whole display */
+        if(LV_LINUX_FBDEV_RENDER_MODE == LV_DISPLAY_RENDER_MODE_DIRECT) {
+            lv_draw_buf_t * draw_buf = lv_display_get_buf_active(disp);
+            src_w = lv_display_get_horizontal_resolution(disp);
+            src_h = lv_display_get_vertical_resolution(disp);
+            src_stride = lv_draw_buf_width_to_stride(src_w, cf);
+            color_p = draw_buf->data;
+            rotated_area.x1 = rotated_area.y1 =  0;
+            lv_area_set_width(&rotated_area, src_w);
+            lv_area_set_height(&rotated_area, src_h);
         }
-
-        /* Rotate the pixel buffer */
-        uint32_t w_stride = lv_draw_buf_width_to_stride(w, cf);
-        uint32_t h_stride = lv_draw_buf_width_to_stride(h, cf);
-
-        switch(rotation) {
-            case LV_DISPLAY_ROTATION_0:
-                break;
-            case LV_DISPLAY_ROTATION_90:
-                lv_draw_sw_rotate(color_p, dsc->rotated_buf, w, h, w_stride, h_stride, rotation, cf);
-                break;
-            case LV_DISPLAY_ROTATION_180:
-                lv_draw_sw_rotate(color_p, dsc->rotated_buf, w, h, w_stride, w_stride, rotation, cf);
-                break;
-            case LV_DISPLAY_ROTATION_270:
-                lv_draw_sw_rotate(color_p, dsc->rotated_buf, w, h, w_stride, h_stride, rotation, cf);
-                break;
+        else {
+            /* For partial and full render modes, we need to rotate the current area
+             * In Full mode we will rotate the whole display just like with direct render mode
+             * but we don't need to do anything special since the area is already the full area of the display
+             * For Partial mode we will rotate just the part we're currently displaying*/
+            src_w = lv_area_get_width(area);
+            src_h = lv_area_get_height(area);
+            src_stride = lv_draw_buf_width_to_stride(lv_area_get_width(area), cf);
+            rotated_area = *area;
         }
-        color_p = dsc->rotated_buf;
-
-        /* Rotate the area */
-        rotated_area = *area;
         lv_display_rotate_area(disp, &rotated_area);
-        area = &rotated_area;
-
-        if(rotation != LV_DISPLAY_ROTATION_180) {
-            w = lv_area_get_width(area);
-            h = lv_area_get_height(area);
+        const uint32_t dest_stride = lv_draw_buf_width_to_stride(lv_area_get_width(&rotated_area), cf);
+        const size_t buf_size = lv_area_get_width(&rotated_area) * lv_area_get_height(&rotated_area) * px_size;
```

Comment:
Not important, just wanted to show this.

```suggestion
        const size_t buf_size = lv_area_get_size(&rotated_area) * px_size;
```

Reply:
That's a good one, I have a follow-up refactor PR for this so I'll do it then